File: tests/test_dispatch_decision.py
"""Tests for dispatch_decision.py schemas."""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "src"))

import unittest
from harness_core.dispatch.dispatch_decision import (
    BudgetReservation,
    DispatchDecision,
    Evidence,
    ExecutionGate,
    RejectedCandidate,
    ShadowRoute,
    COMPLEXITY_WEIGHTS,
    EXECUTION_GATE_TYPES,
    DECISION_STATUSES,
)


def make_evidence(**overrides):
    defaults = dict(
        feature="test_feature",
        text="test text",
        span=(0, 10),
        polarity="positive",
        source="raw_request",
    )
    defaults.update(overrides)
    return Evidence(**defaults)


def make_shadow_route(**overrides):
    defaults = dict(
        tier="balanced_worker",
        profile_id=None,
        reason="test reason",
        admission_scope="diagnostic",
    )
    defaults.update(overrides)
    return ShadowRoute(**defaults)


def make_budget_reservation(**overrides):
    defaults = dict(
        reservation_id="res-001",
        decision_id="dec-001",
        currency="token",
        pre_budget=5000,
        reserved_input_tokens=3000,
        reserved_output_tokens=2000,
        reserved_total_tokens=5000,
        reserved_cost=0.05,
        status="reserved",
        created_at="2026-01-01T00:00:00Z",
        updated_at="2026-01-01T00:00:00Z",
    )
    defaults.update(overrides)
    return BudgetReservation(**defaults)


def make_execution_gate(**overrides):
    defaults = dict(
        gate_id="gate-001",
        gate_type="provider_disabled",
        severity="block",
        reason="provider disabled in Phase 1",
    )
    defaults.update(overrides)
    return ExecutionGate(**defaults)


def make_dispatch_decision(**overrides):
    defaults = dict(
        decision_id="dec-001",
        analysis_id="analysis-001",
        analysis_snapshot={"task_domain": "code", "task_intent": "review"},
        selected_tier="balanced_worker",
        fallback_tier="cheap_executor",
        routing_reason="default policy",
        quality_requirement="standard",
        expected_quality_band="medium",
        confidence=0.8,
        confidence_label="high",
        budget_reservation=make_budget_reservation(),
        execution_policy={"executor_type": "noop", "execution_allowed": True, "requires_human_review": False, "max_retries": 0},
        decision_status="decided",
        created_at="2026-01-01T00:00:00Z",
    )
    defaults.update(overrides)
    return DispatchDecision(**defaults)


class EvidenceTests(unittest.TestCase):
    def test_creation(self):
        e = make_evidence()
        self.assertEqual(e.feature, "test_feature")
        self.assertEqual(e.polarity, "positive")

    def test_to_dict(self):
        d = make_evidence().to_dict()
        self.assertIn("feature", d)
        self.assertIn("span", d)
        self.assertIsInstance(d["span"], list)

    def test_frozen(self):
        e = make_evidence()
        with self.assertRaises(AttributeError):
            e.feature = "changed"


class ShadowRouteTests(unittest.TestCase):
    def test_creation(self):
        sr = make_shadow_route()
        self.assertEqual(sr.admission_scope, "diagnostic")

    def test_to_dict(self):
        d = make_shadow_route().to_dict()
        self.assertIn("tier", d)
        self.assertIn("admission_scope", d)


class BudgetReservationTests(unittest.TestCase):
    def test_creation(self):
        br = make_budget_reservation()
        self.assertEqual(br.status, "reserved")
        self.assertEqual(br.reserved_total_tokens, 5000)

    def test_to_dict(self):
        d = make_budget_reservation().to_dict()
        self.assertIn("reservation_id", d)
        self.assertIn("budget_violation", d)


class ExecutionGateTests(unittest.TestCase):
    def test_creation(self):
        g = make_execution_gate()
        self.assertEqual(g.gate_type, "provider_disabled")
        self.assertFalse(g.cleared)

    def test_to_dict(self):
        d = make_execution_gate().to_dict()
        self.assertIn("gate_id", d)
        self.assertIn("clearance_required", d)


class RejectedCandidateTests(unittest.TestCase):
    def test_creation(self):
        rc = RejectedCandidate(tier="cheap_executor", profile_id=None, reason="too weak")
        self.assertEqual(rc.tier, "cheap_executor")

    def test_to_dict(self):
        d = RejectedCandidate(tier="t", profile_id=None, reason="r").to_dict()
        self.assertIn("tier", d)


class DispatchDecisionTests(unittest.TestCase):
    def test_creation(self):
        dd = make_dispatch_decision()
        self.assertEqual(dd.selected_tier, "balanced_worker")
        self.assertEqual(dd.decision_status, "decided")

    def test_to_dict(self):
        d = make_dispatch_decision().to_dict()
        self.assertIn("decision_id", d)
        self.assertIn("budget_reservation", d)
        self.assertIsInstance(d["budget_reservation"], dict)

    def test_with_gates(self):
        gates = (make_execution_gate(), make_execution_gate(gate_id="g2", gate_type="risk"))
        dd = make_dispatch_decision(execution_gates=gates)
        self.assertEqual(len(dd.execution_gates), 2)

    def test_with_shadow_routes(self):
        routes = (make_shadow_route(), make_shadow_route(tier="cheap_executor"))
        dd = make_dispatch_decision(shadow_routes=routes)
        self.assertEqual(len(dd.shadow_routes), 2)


class ConstantsTests(unittest.TestCase):
    def test_complexity_weights_sum_to_one(self):
        total = sum(COMPLEXITY_WEIGHTS.values())
        self.assertAlmostEqual(total, 1.0, places=6)

    def test_execution_gate_types(self):
        self.assertIn("provider_disabled", EXECUTION_GATE_TYPES)
        self.assertIn("sandbox_disabled", EXECUTION_GATE_TYPES)
        self.assertIn("target_write", EXECUTION_GATE_TYPES)

    def test_decision_statuses(self):
        self.assertIn("decided", DECISION_STATUSES)
        self.assertIn("needs_approval", DECISION_STATUSES)


if __name__ == "__main__":
    unittest.main()

File: engine/tests/test_dispatch_decision.rs
use engine::dispatch_decision::*;

// ---------------------------------------------------------------------------
// EvidenceTests
// ---------------------------------------------------------------------------

#[test]
fn test_evidence_creation() {
    let e = Evidence {
        feature: "test_feature".to_string(),
        polarity: "positive".to_string(),
        ..Default::default()
    };
    assert_eq!(e.feature, "test_feature");
    assert_eq!(e.polarity, "positive");
}

#[test]
fn test_evidence_to_dict() {
    let e = Evidence {
        feature: "f".to_string(),
        text: "t".to_string(),
        span: [0, 5],
        polarity: "positive".to_string(),
        source: "raw_request".to_string(),
        ..Default::default()
    };
    let v = e.to_value();
    assert!(v.get("feature").is_some());
    assert!(v.get("span").is_some());
    assert!(v["span"].is_array());
}

#[test]
fn test_evidence_frozen() {
    let e = Evidence {
        feature: "test_feature".to_string(),
        ..Default::default()
    };
    let _ = e.feature;
}

// ---------------------------------------------------------------------------
// ShadowRouteTests
// ---------------------------------------------------------------------------

#[test]
fn test_shadow_route_creation() {
    let sr = ShadowRoute {
        tier: "balanced_worker".to_string(),
        profile_id: None,
        reason: "test".to_string(),
        admission_scope: "diagnostic".to_string(),
        ..Default::default()
    };
    assert_eq!(sr.admission_scope, "diagnostic");
}

#[test]
fn test_shadow_route_to_dict() {
    let sr = ShadowRoute::default();
    let v = sr.to_value();
    assert!(v.get("tier").is_some());
    assert!(v.get("admission_scope").is_some());
}

// ---------------------------------------------------------------------------
// BudgetReservationTests
// ---------------------------------------------------------------------------

#[test]
fn test_budget_reservation_creation() {
    let br = BudgetReservation {
        status: "reserved".to_string(),
        reserved_total_tokens: 5000,
        ..Default::default()
    };
    assert_eq!(br.status, "reserved");
    assert_eq!(br.reserved_total_tokens, 5000);
}

#[test]
fn test_budget_reservation_to_dict() {
    let br = BudgetReservation::default();
    let v = br.to_value();
    assert!(v.get("reservation_id").is_some());
    assert!(v.get("budget_violation").is_some());
}

// ---------------------------------------------------------------------------
// ExecutionGateTests
// ---------------------------------------------------------------------------

#[test]
fn test_execution_gate_creation() {
    let g = ExecutionGate {
        gate_type: "provider_disabled".to_string(),
        ..Default::default()
    };
    assert_eq!(g.gate_type, "provider_disabled");
    assert!(!g.cleared);
}

#[test]
fn test_execution_gate_to_dict() {
    let g = ExecutionGate::default();
    let v = g.to_value();
    assert!(v.get("gate_id").is_some());
    assert!(v.get("clearance_required").is_some());
}

// ---------------------------------------------------------------------------
// RejectedCandidateTests
// ---------------------------------------------------------------------------

#[test]
fn test_rejected_candidate_creation() {
    let rc = RejectedCandidate {
        tier: "cheap_executor".to_string(),
        profile_id: None,
        reason: "test".to_string(),
        ..Default::default()
    };
    assert_eq!(rc.tier, "cheap_executor");
}

#[test]
fn test_rejected_candidate_to_dict() {
    let rc = RejectedCandidate::default();
    let v = rc.to_value();
    assert!(v.get("tier").is_some());
}

// ---------------------------------------------------------------------------
// DispatchDecisionTests
// ---------------------------------------------------------------------------

#[test]
fn test_dispatch_decision_creation() {
    let dd = DispatchDecision {
        selected_tier: "balanced_worker".to_string(),
        decision_status: "decided".to_string(),
        ..Default::default()
    };
    assert_eq!(dd.selected_tier, "balanced_worker");
    assert_eq!(dd.decision_status, "decided");
}

#[test]
fn test_dispatch_decision_to_dict() {
    let dd = DispatchDecision::default();
    let v = dd.to_value();
    assert!(v.get("decision_id").is_some());
    assert!(v.get("budget_reservation").is_some());
    assert!(v["budget_reservation"].is_object());
}

#[test]
fn test_dispatch_decision_with_gates() {
    let dd = DispatchDecision {
        execution_gates: vec![ExecutionGate::default(), ExecutionGate::default()],
        ..Default::default()
    };
    assert_eq!(dd.execution_gates.len(), 2);
}

#[test]
fn test_dispatch_decision_with_shadow_routes() {
    let dd = DispatchDecision {
        shadow_routes: vec![ShadowRoute::default(), ShadowRoute::default()],
        ..Default::default()
    };
    assert_eq!(dd.shadow_routes.len(), 2);
}

// ---------------------------------------------------------------------------
// ConstantsTests
// ---------------------------------------------------------------------------

#[test]
fn test_complexity_weights_sum_to_one() {
    let weights = complexity_weights();
    let sum: f64 = weights.values().sum();
    assert!((sum - 1.0).abs() < 1e-6, "sum was {}", sum);
}

#[test]
fn test_execution_gate_types() {
    assert!(EXECUTION_GATE_TYPES.contains(&"provider_disabled"));
    assert!(EXECUTION_GATE_TYPES.contains(&"sandbox_disabled"));
    assert!(EXECUTION_GATE_TYPES.contains(&"target_write"));
}

#[test]
fn test_decision_statuses() {
    assert!(DECISION_STATUSES.contains(&"decided"));
    assert!(DECISION_STATUSES.contains(&"needs_approval"));
}
